### ISSUE TITLE
Data objects not connected to binomial, Poisson likelihood when parameters have non-scalar shape

### ISSUE THREAD

**Issue by fonnesbeck** [pymc-devs/pymc#5992]
## Description of your problem

When using `Data` or `MutableData` objects with binomial or Poisson likelihoods, they are not properly added to the model graph when the likelihood has non-scalar shapes. For example, this works fine:

**Please provide a minimal, self-contained, and reproducible example.**
```python
import pymc as pm 
import numpy as np
with pm.Model() as m:
    # Initialize with dummy values
    X = pm.MutableData("X", np.empty(3))

    mu = pm.Exponential('mu', 5)
    theta_like = pm.Poisson(
        "theta_like",
        mu=mu,
        observed=X,
    )
```
![working](https://user-images.githubusercontent.com/81476/180070390-255e7b3e-119b-4bad-8981-cd0dd77c4a0e.svg)

Yet giving `mu` a non-scalar shape breaks the model graph:

```python
import pymc as pm 
import numpy as np
with pm.Model() as m:
    # Initialize with dummy values
    X = pm.MutableData("X", np.empty(3))

    mu = pm.Exponential('mu', 5, shape=3)
    theta_like = pm.Poisson(
        "theta_like",
        mu=mu,
        observed=X,
    )
```

![not_working](https://user-images.githubusercontent.com/81476/180070458-f7f4b7b5-7c9d-48bc-8af0-54e5a813795d.svg)

Interestingly, this does not occur for some other likelihoods (e.g. normal).

I discovered this when the Binomial model appeared not to be learning when sampling with data.

</details>

## Versions and main components

* PyMC/PyMC3 Version: 4.1.2
* Aesara Version:  2.7.5
* Python Version: 3.8.9
* Operating system: Ubuntu
* How did you install PyMC/PyMC3: pip

**Comment by ricardoV94** [comment]
The graphviz is built from dependencies in the generative aesara graph, not in the logp one. X only shows as an input in the generative graph when it is used to define the shape of the likelihood RV. When the parameters specify the same number of dims as observed, the latter isn't used for determining the shape and the arrow doesn't show up.

We should probably refactor the graphviz logic from scratch instead of keep patching it due to old assumptions from V3, over which it was built.

I think it would make more sense to show the Bayesian generative model, where X would make more sense as an output of the likelihood RV.

If we want to plot the logp graph instead, then it would make sense as an input.

**Comment by ricardoV94** [comment]
By the way, this is also indicative of a bug in the variable creation logic, as observed may have broadcasted shapes beyond the parameters (e.g. if `mu.shape=(1,)`), and PyMC would not have taken this into consideration. 

This is related to #5993, although that one shows the bug in relation to dims, and this one is in relation to observed.

**Comment by ricardoV94** [comment]
This was fixed by #6011, it was due to casting of the MutableData. The two graphs now look like this:

![image](https://user-images.githubusercontent.com/28983449/186838106-cf259fa4-a439-470c-b924-9ec93df6e6ec.png)

![image](https://user-images.githubusercontent.com/28983449/186838164-1e6af93d-b644-41f4-af82-81f2345e71ee.png)
